### ISSUE TITLE
after init signal, module cache

### ISSUE THREAD

**Issue by basaran** [awesomeWM/awesome#3378]
Hello,

Is there a signal or mechanism to assign a callback to awesomewm once everything has finished loading? I'm trying to create a global state that depends on the state of other modules, layouts, tags, screens and so on.

Thank you,

**Comment by basaran** [comment]
After further checking into this, I suppose I should be able to use

```
awful.tag.attached_connect_signal(s, "property::selected", function (t)
   ...
end)
```
`t` appears to be pointing to the newly selected tag.

**Comment by Elv13** [comment]
I think what you are looking for is https://awesomewm.org/apidoc/core_components/awesome.html#startup

Another thing you can do is add a delayed call from `gears.timer` directly in `rc.lua`. The delayed call should run after everything started (unless they have their own delayed call). If you are on awesome git-master, there is also more signals in the link above for various other points in the initialization (like when screens are added)

**Comment by basaran** [comment]
Thank you, I'm still learning to get around the api docs, not to mention my lack of lua :)

Is it okay to reload a module through other functions? Would it cause unforseen issues with awesomewm? My global state attempt failed at first, because lua would return the initial value, and not the value I was changing through the signal callback. I suppose this doesn't work like javascript. I did this:

mymodule/something.lua:
```lua
module.regions = {x=4}
return module

awful.tag.attached_connect_signal(s, "property::selected", function (t)
   module.regions = get_regions()
   naughty.notify({text=inspect(module.regions)})
end)
```

mymodule/init.lua
```lua
function reload(m)
  if package.loaded[m] then
    local attrs = {}
    for key,value in pairs(package.loaded[m]) do
      attrs[key] = value
    end
    package.loaded[m] = nil
    temp_module = require(tostring(m))
    for key,value in pairs(attrs) do
      temp_module[key] = value
    end
  else
    temp_module = require(m)
  end
  return temp_module
end

-- some code..
   awful.key({modkey}, "F2", function (c)
      mymodule = reload("mymodule")
      naughty.notify({text=inspect(aw.regions)})
   end), -- client always on top
```

rc.lua
```
require('mymodule')
```
Then the #F2 shortcut, would print the new value.  I'm going to check the `startup` next.

**Comment by Elv13** [comment]
which version of awesome if this?

**Comment by basaran** [comment]
awesome v4.3 (Too long)
 • Compiled against Lua 5.3.5 (running with Lua 5.3)
 • D-Bus support: ✔
 • execinfo support: ✔
 • xcb-randr version: 1.6
 • LGI version: 0.9.2

**Comment by Elv13** [comment]
some tips:

replace

```
awful.tag.attached_connect_signal(s, "property::selected", function (t)
   module.regions = get_regions()
   naughty.notify({text=inspect(module.regions)})
end)
```

with:

```
tag.connect_signal(s, "property::selected", function (t)
   module.regions = get_regions()
   naughty.notify({text=inspect(module.regions)})
end)
```


but more importantly, what are you trying to achieve? Why is reloading a lua module needed? Is there any other way, because that sounds like something that will cause some pain over time. If you are looking for a way to develop faster, there are better ways.

replace `function reload(m)` with `local function`

**Comment by basaran** [comment]
the changes I made produced the same output, I was able to get the updated value that was updated by the callback (I fixed the parameter in tag.connect).

mymodule/something.lua
```lua
module.regions = {x=4}
return module
tag.connect_signal("property::selected", function (t)
   module.regions = get_regions()
   naughty.notify({text=inspect(module.regions)})
end)
```

What didn't work was my first attempt, and it still didn't do with tag.connect, F2 function would print the {x=4}, meanwhile, on reload I could see the correct output from the callback notify. So, it seems the F2 function would not see the new value set by the callback.

mymodule/init.lua
```lua
local something = require('mymodule.something')
.. .. ..
awful.key({modkey}, "F2", function (c)
      naughty.notify({text=inspect(something.regions)})
   end),
```

I'm working on updating my module that I built around layout-machi:

https://github.com/basaran/awesomewm-machina/

The way I have it working right now was to execute my get_regions() functions everytime there is an action and I figured why not cache the whole thing once the tag layout is created and have the keybinds use that :)

And the main reason I checked into this was to come up with a proper way to show fake tabs for those clients in the regions. I had an issue with keeping up with the order of clients (raise, and lower) and that would mess up the fake tab titlebars I was planning to use. I wanted to have a table that I can lookup, and update that I build from layout changes.

I don't know may be I'm just overthinking this while trying to make it efficient.

**Comment by basaran** [comment]
I think I should keep things the way they are without the global state table and finish the implementation. I figured I can assign a custom property `c.region` to the clients and mark them with their region through the existing code and use client signals for that.

Once that's done, I can use the client.get() to keep a stateful order that won't change when I raise and lower clients in a region (at the moment, I'm reading the clients geometrically), and then I won't have to worry about the fake titlebars highlight. I looked into using a wibar but that seemed more complicated and I didn't want to worry about the regions and gaps and placement of things, and updating them when people move clients to different regions. 

Thank you again for making awesomewm, and supporting it.
